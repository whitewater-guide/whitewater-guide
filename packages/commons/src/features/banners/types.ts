import { Overwrite } from 'type-zoo';
import { NamedNode, Node } from '../../core';
import { Group } from '../groups';
import { Region } from '../regions';
import { Connection } from '../types';

export enum BannerPlacement {
  MOBILE_SECTION_DESCRIPTION = 'mobileSectionDescription',
  MOBILE_SECTION_ROW = 'mobileSectionRow',
  MOBILE_SECTION_MEDIA = 'mobileSectionMedia',
  MOBILE_REGION_DESCRIPTION = 'mobileRegionDescription',
}

export enum BannerKind {
  Image = 'Image',
  WebView = 'WebView',
}

export interface BannerSource {
  kind: BannerKind;
  ratio: number | null;
  src: string | null;
}

export interface Banner extends NamedNode {
  enabled: boolean;
  slug: string;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: any;
  // --- connections
  regions?: Connection<Region>;
  groups?: Connection<Group>;
  // ---- transient
  deleted?: boolean;
}

export interface BannerInput {
  id: string | null;
  slug: string;
  name: string;
  enabled: boolean;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: {[key: string]: any} | null;
  // --- connections
  regions: Node[];
  groups: Node[];
}

export type BannerFormInput = Overwrite<BannerInput, { extras: string | null }>;

export function isBanner(node: Node): node is Banner {
  return node.__typename === 'Banner';
}
