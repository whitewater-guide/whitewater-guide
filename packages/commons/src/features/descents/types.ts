import { Timestamped, PageInfo } from '../../apollo';
import { Section } from '../sections';
import { User } from '../users';

export interface DescentLevel {
  __typename?: 'Level';
  unit?: string | null;
  value?: number | null;
}

export interface DescentLevelInput {
  unit?: string | null;
  value: number;
}

export interface Descent extends Timestamped {
  __typename?: 'Descent';
  id: string;
  user: User;
  section: Section;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevel | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentEdge<TCursor = any> {
  __typename?: 'DescentEdge';
  node: Descent;
  cursor: TCursor;
}

export interface DescentInput {
  id?: string | null;

  sectionId: string;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevelInput | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentsConnection<TCursor = any> {
  __typename?: 'DescentsConnection';
  edges: DescentEdge[];
  pageInfo: PageInfo<TCursor>;
}

export interface DescentsFilter {
  startDate?: string | null;
  endDate?: string | null;
  difficulty?: [number, number] | null;
  userId?: string | null;
  sectionId?: string | null;
  sectionName?: string | null;
}
