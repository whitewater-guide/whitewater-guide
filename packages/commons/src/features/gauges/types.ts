import { NamedNode, Node, Timestamped } from '../../core';
import { HarvestStatus, Measurement } from '../measurements';
import { Point, PointInput } from '../points';
import { Source } from '../sources';

export interface Gauge extends NamedNode, Timestamped {
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
  enabled: boolean;
  location: Point;
  source: Source;
  // measurements?: Measurement[];
  lastMeasurement: Measurement | null;
  status: HarvestStatus | null;
}

export interface GaugeInput {
  id: string | null;
  name: string;
  source: Node;
  location: PointInput | null;
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
}
