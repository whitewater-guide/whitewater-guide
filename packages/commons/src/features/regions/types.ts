import {
  Connection,
  NamedNode,
  Node,
  SearchableFilterOptions,
  Timestamped,
} from '../../apollo';
import { Banner } from '../banners';
import { Gauge } from '../gauges';
import { Group } from '../groups';
import { License } from '../licenses';
import { CoordinateLoose, Point, PointInput } from '../points';
import { River } from '../rivers';
import { Section } from '../sections';
import { Source } from '../sources';

export interface Region extends NamedNode, Timestamped {
  description: string | null;
  season: string | null;
  seasonNumeric: number[];
  bounds: Array<[number, number, number]>;
  hidden: boolean | null;
  mapsSize?: number;
  premium: boolean;
  hasPremiumAccess: boolean;
  editable: boolean;
  sku: string | null;
  coverImage: RegionCoverImage;
  pois: Point[];
  groups: Group[];
  copyright: string | null;
  license: License | null;
  // --- connections
  rivers?: Connection<River>;
  gauges?: Connection<Gauge>;
  sections?: Connection<Section>;
  sources?: Connection<Source>;
  mediaSummary?: RegionMediaSummary;
  banners?: Connection<Banner>;
}

export interface RegionInput {
  id: string | null;
  name: string;
  description: string | null;
  copyright: string | null;
  license: License | null;
  season: string | null;
  seasonNumeric: number[];
  bounds: CoordinateLoose[];
  pois: PointInput[];
}

export interface RegionAdminSettings {
  id: string;
  hidden: boolean;
  premium: boolean;
  sku: string | null;
  coverImage: RegionCoverImage;
  mapsSize: number;
}

export interface RegionCoverImage {
  __typename?: string;
  mobile: string | null;
}

export interface RegionMediaSummaryItem {
  count: number;
  size: number;
}

export interface RegionMediaSummary {
  photo: RegionMediaSummaryItem;
  video: RegionMediaSummaryItem;
  blog: RegionMediaSummaryItem;
  maps: RegionMediaSummaryItem;
}

export type RegionFilterOptions = SearchableFilterOptions;

export const DefaultRegionFilterOptions: RegionFilterOptions = {
  searchString: '',
};

export const isRegion = (node?: Node | null): node is Region =>
  !!node && node.__typename === 'Region';
