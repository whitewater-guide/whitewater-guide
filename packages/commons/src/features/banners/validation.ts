import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';

import { BannerKind, BannerPlacement } from './types';

const BannerSourceSchema = yup
  .object({
    kind: yup.mixed().oneOf(Object.values(BannerKind)),
    url: yupTypes.nonEmptyString(),
  })
  .noUnknown();

export const BannerInputSchema = yup
  .object({
    id: yupTypes.uuid().defined().nullable(),
    slug: yupTypes.slug(),
    name: yupTypes.nonEmptyString(),
    priority: yup.number().integer(),
    enabled: yup.bool(),
    placement: yup.mixed().oneOf(Object.values(BannerPlacement)),
    source: BannerSourceSchema.clone(),
    link: yupTypes.https().nullable(true),
    extras: yup.object().nullable(true),
    regions: yup.array(yupTypes.node()),
    groups: yup.array(yupTypes.node()),
  })
  .strict(true)
  .noUnknown(true);
