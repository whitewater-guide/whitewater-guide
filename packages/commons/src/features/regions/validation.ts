// tslint:disable-next-line
import Joi from 'joi';
import { CoordinateSchema, PointInputSchema } from '../points';

export const RegionInputSchema = Joi.object().keys({
  id: Joi.string().guid().allow(null),
  name: Joi.string().min(3).max(100),
  description: Joi.string().allow(null),
  season: Joi.string().allow(null),
  seasonNumeric: Joi.array().max(24).items(Joi.number().min(0).max(23)),
  bounds: Joi.array().items(CoordinateSchema).min(3),
  hidden: Joi.boolean(),
  pois: Joi.array().items(PointInputSchema),
});

// description is draft.js EditorState
export const RegionFormSchema = RegionInputSchema.keys({
  description: Joi.any(),
});
