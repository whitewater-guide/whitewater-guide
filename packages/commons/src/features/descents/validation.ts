import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';
import { DescentLevelInput, DescentInput } from './types';

export const DescentLevelInputSchema = yup
  .object<DescentLevelInput>({
    unit: yup
      .string()
      .max(20)
      .notRequired()
      .nullable(),
    value: yup.number().required(),
  })
  .strict(true)
  .noUnknown();

export const DescentInputSchema = yup
  .object<DescentInput>({
    id: yupTypes
      .uuid()
      .notRequired()
      .nullable(),
    sectionId: yupTypes
      .uuid()
      .required()
      .nullable(false),
    startedAt: yup.date().required() as any,
    duration: yup
      .number()
      .integer()
      .notRequired()
      .nullable(),
    level: DescentLevelInputSchema.clone()
      .notRequired()
      .nullable(),
    comment: yup
      .string()
      .notRequired()
      .nullable(),
    public: yup
      .bool()
      .notRequired()
      .nullable(),
  })
  .strict(true)
  .noUnknown();
