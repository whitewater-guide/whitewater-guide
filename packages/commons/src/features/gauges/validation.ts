import { struct } from '../../utils/validation';
import { PointInputStruct } from '../points';

const GaugeInputFields = {
  id: struct.union(['uuid', 'null']),
  name: 'nonEmptyString',
  code: 'nonEmptyVarchar',
  levelUnit: 'nonEmptyVarchar|null',
  flowUnit: 'nonEmptyVarchar|null',
  location: struct.union([PointInputStruct, 'null']),
  requestParams: 'object|null',
  cron: struct.union(['cron', 'null']),
  url: struct.union(['url', 'null', struct.literal('')]),
  source: 'node',
};

export const GaugeInputStruct = struct.object(GaugeInputFields);

export const GaugeFormStruct = struct.object({
  ...GaugeInputFields,
  requestParams: struct.union(['jsonString', 'null']),
});
