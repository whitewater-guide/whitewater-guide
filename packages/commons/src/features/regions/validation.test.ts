import { createValidator } from '../../utils/validation';
import { RegionAdminSettings, RegionFormInput, RegionInput } from './types';
import {
  RegionAdminSettingsStruct,
  RegionFormStruct,
  RegionInputStruct,
} from './validation';

describe('RegionInput', () => {
  const validator = createValidator(RegionInputStruct);
  type TestValue = [string, RegionInput];

  const correct: RegionInput = {
    id: 'ca844cf2-c3b9-11e8-a355-529269fb1459',
    name: 'region',
    description: 'description',
    bounds: [[1, 1, 1], [2, 2, 2], [3, 3, 3]],
    season: 'when it rains',
    seasonNumeric: [10, 11, 12],
    pois: [
      {
        id: 'b5678064-c34b-11e8-a355-529269fb1459',
        coordinates: [1, 1.3, 200],
        description: 'poi description',
        name: 'some poi',
        kind: 'portage',
      },
    ],
  };

  const correctValues: TestValue[] = [
    ['full value', correct],
    [
      'null value',
      {
        ...correct,
        id: null,
        description: null,
        pois: [],
        season: null,
        seasonNumeric: [],
      },
    ],
    ['empty description', { ...correct, description: '' }],
    ['empty season', { ...correct, season: '' }],
  ];

  const incorrectValues: TestValue[] = [
    ['bad id', { ...correct, id: 'foo' }],
    ['empty name', { ...correct, name: '' }],
    [
      'bad season numeric (bad half-month)',
      { ...correct, seasonNumeric: [10, 100, -10, 11] },
    ],
    [
      'bad season numeric (too many items)',
      { ...correct, seasonNumeric: Array(30).fill(10) },
    ],
    [
      'bad pois',
      {
        ...correct,
        pois: [
          {
            id: null,
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 1, 1],
          },
          {
            id: 'foo',
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 1, 1],
          },
          {
            id: null,
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 2, 1],
          },
        ],
      },
    ],
    ['extra fields', { ...correct, foo: 'bar' } as any],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    expect(validator(value)).not.toBeNull();
    expect(validator(value)).toMatchSnapshot();
  });
});

describe('RegionFormInput', () => {
  interface TestRichText {
    markdown: string;
    prosemirror: { state: string };
  }
  type TestFormInput = RegionFormInput<TestRichText>;
  const validator = createValidator(RegionFormStruct());
  type TestValue = [string, TestFormInput];

  const correct: TestFormInput = {
    id: 'ca844cf2-c3b9-11e8-a355-529269fb1459',
    name: 'region',
    description: {
      markdown: 'markdown',
      prosemirror: { state: 'prosemirror' },
    },
    bounds: [[1, 1, 1], [2, 2, 2], [3, 3, 3]],
    season: 'when it rains',
    seasonNumeric: [10, 11, 12],
    pois: [
      {
        id: 'b5678064-c34b-11e8-a355-529269fb1459',
        coordinates: [1, 1.3, 200],
        description: 'poi description',
        name: 'some poi',
        kind: 'portage',
      },
    ],
  };

  const correctValues: TestValue[] = [
    ['full value', correct],
    [
      'null value',
      {
        ...correct,
        id: null,
        pois: [],
        season: null,
        seasonNumeric: [],
      },
    ],
    ['empty season', { ...correct, season: '' }],
  ];

  const incorrectValues: TestValue[] = [
    ['bad id', { ...correct, id: 'foo' }],
    ['empty name', { ...correct, name: '' }],
    [
      'bad season numeric (bad half-month)',
      { ...correct, seasonNumeric: [10, 100, -10, 11] },
    ],
    [
      'bad season numeric (too many items)',
      { ...correct, seasonNumeric: Array(30).fill(10) },
    ],
    [
      'bad pois',
      {
        ...correct,
        pois: [
          {
            id: null,
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 1, 1],
          },
          {
            id: 'foo',
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 1, 1],
          },
          {
            id: null,
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 2, 1],
          },
        ],
      },
    ],
    ['extra fields', { ...correct, foo: 'bar' } as any],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    expect(validator(value)).not.toBeNull();
    expect(validator(value)).toMatchSnapshot();
  });
});

describe('RegionAdminSettings', () => {
  const validator = createValidator(RegionAdminSettingsStruct);
  type TestValue = [string, RegionAdminSettings];

  const correct: RegionAdminSettings = {
    id: '4e0d6c92-c3bb-11e8-a355-529269fb1459',
    hidden: false,
    premium: false,
    sku: 'region.sku',
    coverImage: {
      mobile: 'cover.jpg',
    },
  };

  const correctValues: TestValue[] = [
    ['full value', correct],
    [
      'null value',
      {
        ...correct,
        sku: null,
        coverImage: {
          mobile: null,
        },
      },
    ],
  ];

  const incorrectValues: TestValue[] = [
    ['bad id', { ...correct, id: 'foo' }],
    ['empty sku', { ...correct, sku: '' }],
    ['bad sku 1', { ...correct, sku: 'region.x' }],
    ['bad sku 2', { ...correct, sku: 'region.g@licia' }],
    ['empty cover image', { ...correct, coverImage: { mobile: '' } }],
    ['extra fields', { ...correct, foo: 'bar' } as any],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    expect(validator(value)).not.toBeNull();
    expect(validator(value)).toMatchSnapshot();
  });
});
