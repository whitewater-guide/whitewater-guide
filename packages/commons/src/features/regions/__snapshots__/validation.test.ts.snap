// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegionAdminSettings should be invalid for bad id 1`] = `
Object {
  "id": "Incorrect UUID",
}
`;

exports[`RegionAdminSettings should be invalid for bad sku 1 1`] = `
Object {
  "sku": "SKU must start with \\"region.\\" and contain at least 3 word characters after dot",
}
`;

exports[`RegionAdminSettings should be invalid for bad sku 2 1`] = `
Object {
  "sku": "SKU must start with \\"region.\\" and contain at least 3 word characters after dot",
}
`;

exports[`RegionAdminSettings should be invalid for empty banner 1`] = `
Object {
  "banners": Object {
    "regionDescriptionMobile": "Value must be non-empty string",
  },
}
`;

exports[`RegionAdminSettings should be invalid for empty cover image 1`] = `
Object {
  "coverImage": Object {
    "mobile": "Value must be non-empty string",
  },
}
`;

exports[`RegionAdminSettings should be invalid for empty sku 1`] = `
Object {
  "sku": "SKU must start with \\"region.\\" and contain at least 3 word characters after dot",
}
`;

exports[`RegionAdminSettings should be invalid for extra fields 1`] = `
Object {
  "foo": "Incorrect value",
}
`;

exports[`RegionFormInput should be invalid for bad id 1`] = `
Object {
  "id": "Incorrect UUID",
}
`;

exports[`RegionFormInput should be invalid for bad pois 1`] = `
Object {
  "pois": Array [
    undefined,
    Object {
      "id": "Incorrect UUID",
    },
  ],
}
`;

exports[`RegionFormInput should be invalid for bad season numeric (bad half-month) 1`] = `
Object {
  "seasonNumeric": Array [
    undefined,
    "Must be integer between 0 and 23",
    "Must be integer between 0 and 23",
  ],
}
`;

exports[`RegionFormInput should be invalid for bad season numeric (too many items) 1`] = `
Object {
  "seasonNumeric": "Maximal length is 24",
}
`;

exports[`RegionFormInput should be invalid for empty name 1`] = `
Object {
  "name": "Value must be non-empty string",
}
`;

exports[`RegionFormInput should be invalid for extra fields 1`] = `
Object {
  "foo": "Incorrect value",
}
`;

exports[`RegionInput should be invalid for bad id 1`] = `
Object {
  "id": "Incorrect UUID",
}
`;

exports[`RegionInput should be invalid for bad pois 1`] = `
Object {
  "pois": Array [
    undefined,
    Object {
      "id": "Incorrect UUID",
    },
  ],
}
`;

exports[`RegionInput should be invalid for bad season numeric (bad half-month) 1`] = `
Object {
  "seasonNumeric": Array [
    undefined,
    "Must be integer between 0 and 23",
    "Must be integer between 0 and 23",
  ],
}
`;

exports[`RegionInput should be invalid for bad season numeric (too many items) 1`] = `
Object {
  "seasonNumeric": "Maximal length is 24",
}
`;

exports[`RegionInput should be invalid for empty name 1`] = `
Object {
  "name": "Value must be non-empty string",
}
`;

exports[`RegionInput should be invalid for extra fields 1`] = `
Object {
  "foo": "Incorrect value",
}
`;
